File: src/pulseaudio/pulseaudio.cpp
#include "pulseaudio/pulseaudio.h"

#include <spdlog/spdlog.h>

#include <algorithm>
#include <chrono>
#include <cstring>
#include <set>
#include <thread>

#ifdef HAVE_PULSEAUDIO
#include <pulse/pulseaudio.h>
#endif

namespace autowhisper {

struct MutedSinkInput {
    uint32_t index;
    std::string name;
    bool was_muted;
};

struct PulseAudioManager::Impl {
#ifdef HAVE_PULSEAUDIO
    pa_mainloop* mainloop = nullptr;
    pa_context* context = nullptr;
#endif
    std::vector<MutedSinkInput> muted_inputs;
    std::thread mute_timer;
    bool mute_timer_active = false;

    static const std::set<std::string> EXCLUDED_APPS;
};

const std::set<std::string> PulseAudioManager::Impl::EXCLUDED_APPS = {
    "python", "python3", "autowhisper", "sounddevice"
};

PulseAudioManager::PulseAudioManager(bool enabled, float beep_duration)
    : enabled_(enabled), beep_duration_(beep_duration), impl_(std::make_unique<Impl>()) {}

PulseAudioManager::~PulseAudioManager() {
    cleanup();
}

#ifdef HAVE_PULSEAUDIO

namespace {

// Simple blocking iteration helper
void iterate(pa_mainloop* ml, pa_operation* op) {
    if (!op) return;
    while (pa_operation_get_state(op) == PA_OPERATION_RUNNING) {
        pa_mainloop_iterate(ml, 1, nullptr);
    }
    pa_operation_unref(op);
}

struct SinkInputListData {
    pa_mainloop* ml;
    std::vector<MutedSinkInput>* inputs_to_mute;
    const std::set<std::string>* excluded;
};

void sink_input_list_cb(pa_context* /*ctx*/, const pa_sink_input_info* info, int eol, void* userdata) {
    if (eol > 0 || !info) return;

    auto* data = static_cast<SinkInputListData*>(userdata);

    // Get binary name
    const char* binary = pa_proplist_gets(info->proplist, PA_PROP_APPLICATION_PROCESS_BINARY);
    std::string binary_name = binary ? binary : "";

    // Extract basename
    auto pos = binary_name.rfind('/');
    if (pos != std::string::npos) {
        binary_name = binary_name.substr(pos + 1);
    }

    // Check exclusion
    if (data->excluded->count(binary_name)) {
        spdlog::debug("Excluding '{}' from muting (binary: {})", info->name, binary_name);
        return;
    }

    MutedSinkInput muted;
    muted.index = info->index;
    muted.name = info->name ? info->name : "Unknown";
    muted.was_muted = info->mute != 0;
    data->inputs_to_mute->push_back(muted);
}

} // anonymous namespace

bool PulseAudioManager::initialize() {
    if (!enabled_) {
        spdlog::debug("PulseAudio muting disabled by configuration");
        return true;
    }

    impl_->mainloop = pa_mainloop_new();
    if (!impl_->mainloop) {
        spdlog::warn("Failed to create PulseAudio mainloop. Muting disabled.");
        enabled_ = false;
        return true;
    }

    pa_mainloop_api* api = pa_mainloop_get_api(impl_->mainloop);
    impl_->context = pa_context_new(api, "autowhisper");
    if (!impl_->context) {
        spdlog::warn("Failed to create PulseAudio context. Muting disabled.");
        pa_mainloop_free(impl_->mainloop);
        impl_->mainloop = nullptr;
        enabled_ = false;
        return true;
    }

    pa_context_connect(impl_->context, nullptr, PA_CONTEXT_NOFLAGS, nullptr);

    // Wait for connection
    while (true) {
        pa_mainloop_iterate(impl_->mainloop, 1, nullptr);
        auto state = pa_context_get_state(impl_->context);
        if (state == PA_CONTEXT_READY) break;
        if (state == PA_CONTEXT_FAILED || state == PA_CONTEXT_TERMINATED) {
            spdlog::warn("PulseAudio connection failed. Muting disabled.");
            pa_context_unref(impl_->context);
            pa_mainloop_free(impl_->mainloop);
            impl_->context = nullptr;
            impl_->mainloop = nullptr;
            enabled_ = false;
            return true;
        }
    }

    initialized_ = true;
    spdlog::info("PulseAudio muting enabled");
    return true;
}

void PulseAudioManager::do_mute() {
    if (!enabled_ || !initialized_ || !impl_->context) return;

    std::lock_guard<std::mutex> lock(lock_);

    // List sink inputs
    std::vector<MutedSinkInput> to_mute;
    SinkInputListData data{impl_->mainloop, &to_mute, &Impl::EXCLUDED_APPS};

    auto* op = pa_context_get_sink_input_info_list(impl_->context, sink_input_list_cb, &data);
    iterate(impl_->mainloop, op);

    int muted_count = 0;
    for (auto& si : to_mute) {
        impl_->muted_inputs.push_back(si);
        if (!si.was_muted) {
            auto* mop = pa_context_set_sink_input_mute(impl_->context, si.index, 1, nullptr, nullptr);
            iterate(impl_->mainloop, mop);
            muted_count++;
            spdlog::debug("Muted: {}", si.name);
        }
    }

    if (muted_count > 0) {
        spdlog::debug("Muted {} other app(s)", muted_count);
    }
}

void PulseAudioManager::mute_other_apps(bool delay) {
    if (!enabled_ || !initialized_) return;

    // Cancel pending timer
    if (impl_->mute_timer_active && impl_->mute_timer.joinable()) {
        impl_->mute_timer_active = false;
        impl_->mute_timer.join();
    }

    if (delay) {
        float delay_seconds = beep_duration_ + 0.05f;
        spdlog::debug("Scheduling app mute in {:.2f}s", delay_seconds);
        impl_->mute_timer_active = true;
        impl_->mute_timer = std::thread([this, delay_seconds]() {
            std::this_thread::sleep_for(
                std::chrono::milliseconds(static_cast<int>(delay_seconds * 1000)));
            if (impl_->mute_timer_active) {
                do_mute();
            }
            impl_->mute_timer_active = false;
        });
    } else {
        do_mute();
    }
}

void PulseAudioManager::unmute_other_apps() {
    if (!enabled_ || !initialized_ || !impl_->context) return;

    // Cancel pending mute
    if (impl_->mute_timer_active) {
        impl_->mute_timer_active = false;
        if (impl_->mute_timer.joinable()) impl_->mute_timer.join();
    }

    std::lock_guard<std::mutex> lock(lock_);

    if (impl_->muted_inputs.empty()) return;

    int restored = 0;
    for (const auto& si : impl_->muted_inputs) {
        if (!si.was_muted) {
            auto* op = pa_context_set_sink_input_mute(impl_->context, si.index, 0, nullptr, nullptr);
            if (op) {
                iterate(impl_->mainloop, op);
                restored++;
                spdlog::debug("Unmuted: {}", si.name);
            }
        }
    }

    if (restored > 0) {
        spdlog::debug("Restored audio for {} app(s)", restored);
    }

    impl_->muted_inputs.clear();
}

void PulseAudioManager::cleanup() {
    if (impl_->mute_timer_active) {
        impl_->mute_timer_active = false;
        if (impl_->mute_timer.joinable()) impl_->mute_timer.join();
    }

    unmute_other_apps();

    if (impl_->context) {
        pa_context_disconnect(impl_->context);
        pa_context_unref(impl_->context);
        impl_->context = nullptr;
    }
    if (impl_->mainloop) {
        pa_mainloop_free(impl_->mainloop);
        impl_->mainloop = nullptr;
    }

    initialized_ = false;
    spdlog::debug("PulseAudio manager cleaned up");
}

#else // !HAVE_PULSEAUDIO

bool PulseAudioManager::initialize() {
    if (enabled_) {
        spdlog::warn("PulseAudio support not compiled in. Muting disabled.");
        enabled_ = false;
    }
    return true;
}

void PulseAudioManager::do_mute() {}
void PulseAudioManager::mute_other_apps(bool /*delay*/) {}
void PulseAudioManager::unmute_other_apps() {}
void PulseAudioManager::cleanup() { spdlog::debug("PulseAudio manager cleaned up (stub)"); }

#endif // HAVE_PULSEAUDIO

} // namespace autowhisper

File: src/autowhisper/pulseaudio.py
"""PulseAudio manager for muting other applications during recording."""

import logging
import os
import threading
from dataclasses import dataclass

logger = logging.getLogger(__name__)

# Try to import pulsectl, but gracefully handle if not available
try:
    import pulsectl
    PULSECTL_AVAILABLE = True
except ImportError:
    pulsectl = None
    PULSECTL_AVAILABLE = False


@dataclass
class MutedSinkInput:
    """Represents a muted sink input with its original state."""
    index: int
    name: str
    was_muted: bool  # Original state to restore


class PulseAudioManager:
    """Manages muting/unmuting of other applications via PulseAudio."""

    EXCLUDED_APPS = {"python", "python3", "autowhisper", "sounddevice"}

    def __init__(self, enabled: bool, beep_duration: float):
        """
        Initialize the PulseAudio manager.

        Args:
            enabled: Whether app muting feature is enabled
            beep_duration: Duration of the start beep in seconds
        """
        self._enabled = enabled
        self._beep_duration = beep_duration
        self._pulse: pulsectl.Pulse | None = None
        self._muted_inputs: list[MutedSinkInput] = []
        self._lock = threading.Lock()
        self._mute_timer: threading.Timer | None = None
        self._initialized = False

    def initialize(self) -> bool:
        """
        Initialize the PulseAudio connection.

        Returns:
            True if initialization succeeded or feature is disabled/unavailable
            (graceful degradation), False only on catastrophic errors.
        """
        if not self._enabled:
            logger.debug("PulseAudio muting disabled by configuration")
            return True

        if not PULSECTL_AVAILABLE:
            logger.warning(
                "pulsectl not available, app muting feature disabled. "
                "Install with: pip install pulsectl"
            )
            self._enabled = False
            return True

        try:
            self._pulse = pulsectl.Pulse("autowhisper")
            self._initialized = True
            logger.info("PulseAudio muting enabled")
            return True
        except Exception as e:
            logger.warning(f"Failed to connect to PulseAudio: {e}. App muting disabled.")
            self._enabled = False
            return True

    def _is_excluded(self, sink_input) -> bool:
        """
        Check if a sink input should be excluded from muting.

        Args:
            sink_input: PulseAudio sink input object

        Returns:
            True if the sink input should NOT be muted
        """
        proplist = sink_input.proplist
        binary = proplist.get("application.process.binary", "")

        # Get the basename (filename without path)
        basename = os.path.basename(binary)

        if basename in self.EXCLUDED_APPS:
            logger.debug(f"Excluding sink input '{sink_input.name}' (binary: {basename})")
            return True

        return False

    def _do_mute(self) -> None:
        """Actually perform the muting of other apps (called after delay)."""
        if not self._enabled or not self._initialized or not self._pulse:
            return

        with self._lock:
            try:
                sink_inputs = self._pulse.sink_input_list()
                muted_count = 0

                for sink_input in sink_inputs:
                    if self._is_excluded(sink_input):
                        continue

                    # Track original mute state
                    was_muted = sink_input.mute == 1
                    self._muted_inputs.append(
                        MutedSinkInput(
                            index=sink_input.index,
                            name=sink_input.name or "Unknown",
                            was_muted=was_muted,
                        )
                    )

                    # Only mute if not already muted
                    if not was_muted:
                        self._pulse.sink_input_mute(sink_input.index, mute=True)
                        muted_count += 1
                        logger.debug(f"Muted: {sink_input.name}")

                if muted_count > 0:
                    logger.debug(f"Muted {muted_count} other app(s)")

            except Exception as e:
                logger.warning(f"Error muting other apps: {e}")
                # Clear tracked inputs on error to avoid issues during unmute
                self._muted_inputs.clear()

    def mute_other_apps(self, delay: bool = True) -> None:
        """
        Mute all other applications' audio.

        Args:
            delay: If True, wait for beep_duration + 0.05s before muting
                   so user hears the start beep
        """
        if not self._enabled or not self._initialized:
            return

        # Cancel any pending mute timer
        if self._mute_timer is not None:
            self._mute_timer.cancel()
            self._mute_timer = None

        if delay:
            # Schedule muting after beep finishes
            delay_seconds = self._beep_duration + 0.05
            logger.debug(f"Scheduling app mute in {delay_seconds:.2f}s")
            self._mute_timer = threading.Timer(delay_seconds, self._do_mute)
            self._mute_timer.daemon = True
            self._mute_timer.start()
        else:
            self._do_mute()

    def unmute_other_apps(self) -> None:
        """Restore original mute state for all previously muted apps."""
        if not self._enabled or not self._initialized or not self._pulse:
            return

        # Cancel any pending mute timer (in case unmute called before mute executed)
        if self._mute_timer is not None:
            self._mute_timer.cancel()
            self._mute_timer = None

        with self._lock:
            if not self._muted_inputs:
                return

            restored_count = 0

            for muted_input in self._muted_inputs:
                # Only unmute apps that weren't muted before we muted them
                if not muted_input.was_muted:
                    try:
                        self._pulse.sink_input_mute(muted_input.index, mute=False)
                        restored_count += 1
                        logger.debug(f"Unmuted: {muted_input.name}")
                    except Exception as e:
                        # Sink input may no longer exist (app closed)
                        logger.debug(
                            f"Could not unmute {muted_input.name} "
                            f"(index {muted_input.index}): {e}"
                        )

            if restored_count > 0:
                logger.debug(f"Restored audio for {restored_count} app(s)")

            self._muted_inputs.clear()

    def cleanup(self) -> None:
        """Clean up resources on shutdown."""
        # Cancel any pending mute timer
        if self._mute_timer is not None:
            self._mute_timer.cancel()
            self._mute_timer = None

        # Restore any muted apps
        self.unmute_other_apps()

        # Close PulseAudio connection
        if self._pulse is not None:
            try:
                self._pulse.close()
            except Exception as e:
                logger.debug(f"Error closing PulseAudio connection: {e}")
            self._pulse = None

        self._initialized = False
        logger.debug("PulseAudio manager cleaned up")
